### PR TITLE
feat(hmr-plugin): add option for persisting state after the root module is disposed

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: https://github.com/ngxs/store/issues/1347

## What is the new behavior?

If during the method of ngOnDestroy the state is deleted before it is saved, we can save the state again after.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/hmr-plugin/src/symbols.ts [hunk 1]
```diff
@@ -37,6 +37,12 @@ export interface NgxsHmrOptions {
    * (default: 100ms)
    */
   deferTime?: number;
+
+  /**
+   * @description
+   * additionally persist state after destruction AppModule
+   */
+  persistAfterDestroyed?: boolean;
```
Comment (arturovt) [inline packages/hmr-plugin/src/symbols.ts]:
```suggestion
  persistAfterDestroy?: boolean;
```

File: packages/hmr-plugin/src/symbols.ts [hunk 2]
```diff
@@ -37,6 +37,12 @@ export interface NgxsHmrOptions {
    * (default: 100ms)
    */
   deferTime?: number;
+
+  /**
+   * @description
+   * additionally persist state after destruction AppModule
```
Comment (arturovt) [inline packages/hmr-plugin/src/symbols.ts]:
```suggestion
   * Additionally persist state when the AppModule is disposed
```

File: packages/hmr-plugin/src/tests/hmr-manager.spec.ts [hunk 3]
```diff
@@ -290,4 +291,86 @@ describe('HMR Plugin', () => {
     webpackModule.destroyModule();
     expect((appModule as any)['_destroyed']).toEqual(true);
   });
+
+  describe('side effect destroy state when called reset in ngOnDestroy', () => {
+    it('state provided before module destroy', fakeAsync(async () => {
```
Comment (arturovt) [inline packages/hmr-plugin/src/tests/hmr-manager.spec.ts]:
```suggestion
    it('state has to be provided before module is disposed', fakeAsync(async () => {
```

File: packages/hmr-plugin/src/tests/hmr-manager.spec.ts [hunk 4]
```diff
@@ -290,4 +291,86 @@ describe('HMR Plugin', () => {
     webpackModule.destroyModule();
     expect((appModule as any)['_destroyed']).toEqual(true);
   });
+
+  describe('side effect destroy state when called reset in ngOnDestroy', () => {
```
Comment (arturovt) [inline packages/hmr-plugin/src/tests/hmr-manager.spec.ts]: What do you mean here? Probably lost in translation :wink:

File: packages/hmr-plugin/src/tests/hmr-manager.spec.ts [hunk 5]
```diff
@@ -290,4 +291,86 @@ describe('HMR Plugin', () => {
...
+      });
+
+      webpackModule.destroyModule();
+
+      tick(1000);
+
+      expect(store.snapshot()).toEqual({});
+    }));
+
+    it('state provided after module destroy', fakeAsync(async () => {
```
Comment (arturovt) [inline packages/hmr-plugin/src/tests/hmr-manager.spec.ts]:
```suggestion
    it('state has to be provided after module is disposed', fakeAsync(async () => {
```

File: packages/hmr-plugin/src/tests/hmr-manager.spec.ts [hunk 6]
```diff
@@ -290,4 +291,86 @@ describe('HMR Plugin', () => {
...
+      webpackModule.destroyModule();
+
+      tick(1000);
+
+      expect(store.snapshot()).toEqual({
+        mock_state: { value: 'test' }
+      });
+    }));
+
+    it('state provided before module destroy with reset in ngOnDestroy', fakeAsync(async () => {
```
Comment (arturovt) [inline packages/hmr-plugin/src/tests/hmr-manager.spec.ts]:
```suggestion
    it('state has to be provided before module is disposed with calling reset in ngOnDestroy', fakeAsync(async () => {
```

**Comment by splincode** [comment]
@arturovt done

File: docs/plugins/hmr.md [hunk 7]
```diff
@@ -180,3 +180,23 @@ export class MyState {
   }
 }
 ```
+
+### HMR Options
+
+The following options are available:
+
+- `autoClearLogs` - Clear log after each hmr update (default value is `true`).
```
Comment (arturovt) [inline docs/plugins/hmr.md]:
```suggestion
- `autoClearLogs` - Clear logs after each refresh (default value is `true`).
```

Comment (splincode) [reply docs/plugins/hmr.md]: done

**Pull request merged**